SVG textLength not working correctly

Ahmad-S792 opened the pull request:
#### 5c38a23d96c2b79d8dbab14b7a7926eac78d4001
<pre>
SVG textLength not working correctly

SVG textLength not working correctly
<a href="https://bugs.webkit.org/show_bug.cgi?id=171805">https://bugs.webkit.org/show_bug.cgi?id=171805</a>

Reviewed by Simon Fraser.

Merge - <a href="https://chromium.googlesource.com/chromium/src/+/308e792b8d4648ba4e298c80faca80bc780b5735">https://chromium.googlesource.com/chromium/src/+/308e792b8d4648ba4e298c80faca80bc780b5735</a> &amp;
<a href="https://chromium.googlesource.com/chromium/src/+/1cc74004b24c8a79dcee523e913c1d929c1167da">https://chromium.googlesource.com/chromium/src/+/1cc74004b24c8a79dcee523e913c1d929c1167da</a>

There should be no advance adjustment after the last character (really,
&quot;typographic unit&quot;), so divide the excess space by the character count
minus one - i.e the number of gaps - rather than the character count.

Further, extend it to adjusting the &apos;textPath&apos;, &apos;textLength&apos; computation in the same way.

* Source/WebCore/rendering/svg/SVGTextChunk.cpp:
(SVGTextChunk::layout): Update to account for last character in adjustment
* Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
(SVGTextLayoutEngine::beginTextPathLayout): Ditto
* LayoutTests/imported/w3c/web-platform-tests/svg/text/scripted/texpath-textLength-text-anchor-001.tentative-expected.txt: Rebaselined
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-text-01-b-manual-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textLayout-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textLayout-expected.png: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textLength-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textLength-expected.png: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textOnPathSpaces-expected.png: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textOnPathSpaces-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-1-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-1-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-2-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-2-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-4-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-4-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-1-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-1-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-2-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-2-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-3-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-3-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-4-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-4-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/text-text-01-b-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-text-01-b-expected.png: Ditto
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/mac/svg/batik/text/textLayout-expected.png: Ditto
* LayoutTests/platform/mac/svg/batik/text/textLayout-expected.txt: Ditto
* LayoutTests/platform/mac/svg/batik/text/texOnPathSpaces-expected.txt: Ditto
* LayoutTests/platform/mac/svg/batik/text/texOnPathSpaces-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-1-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-1-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-2-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-2-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-3-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-3-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-4-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-4-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-1-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-1-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-2-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-2-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-3-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-3-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-4-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-4-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-text-01-b-expected.png: Ditto
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-text-01-b-expected.png: Ditto
* LayoutTests/platform/ios/svg/batik/text/textLayout-expected.txt: Ditto
* LayoutTests/platform/ios/svg/batik/text/textLength-expected.txt: Ditto
* LayoutTests/platform/ios/svg/batik/text/textOnPathSpaces-expected.txt: Ditto
* LayoutTests/platform/ios/svg/text/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/text-text-01-b-manual-expected.txt: Ditto
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/text-text-01-b-manual-expected.txt: Ditto

Canonical link: <a href="https://commits.webkit.org/258921@main">https://commits.webkit.org/258921@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/678a7de619d60658f431858726b8bac11cb1f757

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/103348 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/12469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/36313 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/112584 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/172788 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/13498 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/3374 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/95565 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/111337 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/109123 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/10371 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/93484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/38001 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/92191 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/25036 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/79731 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/5860 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/26439 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6033 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/2969 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12015 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/45948 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/6132 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/7791 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
```diff
@@ -206,9 +206,10 @@ void SVGTextLayoutEngine::beginTextPathLayout(RenderSVGTextPath& textPath, SVGTe
-    else
+    if (textContentElement->lengthAdjust() == SVGLengthAdjustSpacing) {
+        if (totalCharacters > 1)
+        m_textPathSpacing = (desiredTextLength - totalLength) / (totalCharacters - 1);
```
Wrong indentation here.

Ahmad-S792 replied:
Will fix it. Waiting on EWS run of if any failing tests. Once all pass, will push new with indentation fix and then will merge. Thanks for review.

Ahmad-S792 replied:
Fixed.

Ahmad-S792 committed:
SVG textLength not working correctly

SVG textLength not working correctly
https://bugs.webkit.org/show_bug.cgi?id=171805

Reviewed by Simon Fraser.

Merge - https://chromium.googlesource.com/chromium/src/+/308e792b8d4648ba4e298c80faca80bc780b5735 &
https://chromium.googlesource.com/chromium/src/+/1cc74004b24c8a79dcee523e913c1d929c1167da

There should be no advance adjustment after the last character (really,
"typographic unit"), so divide the excess space by the character count
minus one - i.e the number of gaps - rather than the character count.

Further, extend it to adjusting the 'textPath', 'textLength' computation in the same way.

* Source/WebCore/rendering/svg/SVGTextChunk.cpp:
(SVGTextChunk::layout): Update to account for last character in adjustment
* Source/WebCore/rendering/svg/SVGTextLayoutEngine.cpp:
(SVGTextLayoutEngine::beginTextPathLayout): Ditto
* LayoutTests/imported/w3c/web-platform-tests/svg/text/scripted/texpath-textLength-text-anchor-001.tentative-expected.txt: Rebaselined
* LayoutTests/platform/gtk/imported/w3c/web-platform-tests/svg/import/text-text-01-b-manual-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textLayout-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textLayout-expected.png: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textLength-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textLength-expected.png: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textOnPathSpaces-expected.png: Ditto
* LayoutTests/platform/gtk/svg/batik/text/textOnPathSpaces-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-1-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-1-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-2-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-2-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-4-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-squeeze-4-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-1-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-1-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-2-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-2-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-3-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-3-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-4-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/text/select-textLength-spacing-stretch-4-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/text-text-01-b-expected.png: Ditto
* LayoutTests/platform/gtk/svg/text/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-text-01-b-expected.png: Ditto
* LayoutTests/platform/gtk/svg/W3C-SVG-1.1/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/mac/svg/batik/text/textLayout-expected.png: Ditto
* LayoutTests/platform/mac/svg/batik/text/textLayout-expected.txt: Ditto
* LayoutTests/platform/mac/svg/batik/text/texOnPathSpaces-expected.txt: Ditto
* LayoutTests/platform/mac/svg/batik/text/texOnPathSpaces-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-1-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-1-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-2-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-2-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-3-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-3-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-4-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-squeeze-4-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-1-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-1-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-2-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-2-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-3-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-3-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-4-expected.png: Ditto
* LayoutTests/platform/mac/svg/text/select-textLength-spacing-stretch-4-expected.txt: Ditto
* LayoutTests/platform/mac/svg/text/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-text-01-b-expected.png: Ditto
* LayoutTests/platform/mac/svg/W3C-SVG-1.1/text-text-01-b-expected.png: Ditto
* LayoutTests/platform/ios/svg/batik/text/textLayout-expected.txt: Ditto
* LayoutTests/platform/ios/svg/batik/text/textLength-expected.txt: Ditto
* LayoutTests/platform/ios/svg/batik/text/textOnPathSpaces-expected.txt: Ditto
* LayoutTests/platform/ios/svg/text/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/ios/svg/W3C-SVG-1.1/text-text-01-b-expected.txt: Ditto
* LayoutTests/platform/ios-simulator/imported/w3c/web-platform-tests/svg/import/text-text-01-b-manual-expected.txt: Ditto
* LayoutTests/platform/mac/imported/w3c/web-platform-tests/svg/import/text-text-01-b-manual-expected.txt: Ditto

Canonical link: https://commits.webkit.org/258921@main
